include/envoy/ssl/private_key/private_key_callbacks.h

```diff
@@ -0,0 +1,30 @@
+#pragma once
+
+#include <functional>
+#include <string>
+
+#include "envoy/common/pure.h"
+
+namespace Envoy {
+namespace Ssl {
+
+enum class PrivateKeyOperationStatus {
+  Success,
+  Failure,
+};
+
+class PrivateKeyOperationsCallbacks {
+public:
+  virtual ~PrivateKeyOperationsCallbacks() {}
+
+  /**
+   * Callback function which is called when the asynchronous private key
+   * operation has been completed.
+   * @param status is "Success" or "Failure" depending on whether the private key operation was
+   * successful or not.
+   */
+  virtual void complete(PrivateKeyOperationStatus status) PURE;
```

Comment:
I'm a bit worried about the case where the connection (and `SslSocket`) goes away while waiting for the async handshake to complete, and then the `complete()` callback is called, resulting in crash.

Do we have any protections against that?

Reply:
Originally the idea was that the `PrivateKeyOperations` lifecycle would be tied to the `SslSocket` lifecycle. I'll have to do this bit differently now when the operations are tied to certificates (which may change on-the-fly if multiple certificate support is used).